aesd-char-driver/main.c

```diff
@@ -220,18 +222,152 @@ ssize_t aesd_write(struct file *filp, const char __user *ubuf, size_t count, lof
         /* Loop in case multiple '\n' exist in the now-updated incomplete_cmd */
     }
 
+   
+
   out_unlock_free:
+    if (retval > 0)                 // only advance when we actually wrote bytes
+        *f_pos += retval;           // keep positional I/O consistent with llseek
     mutex_unlock(&dev->lock);
     kfree(kbuf);
     return retval;
 }
+
+static loff_t aesd_buffer_size_bytes(const struct aesd_circular_buffer *buf)
+{
+    loff_t total = 0;
+    uint8_t idx;
+    struct aesd_buffer_entry *entry;
+
+    AESD_CIRCULAR_BUFFER_FOREACH(entry, buf, idx) {
+        if (entry && entry->buffptr) 
+          total += entry->size;
+    }
+    return total;
+}
+/* ---------- llseek implementation---------- */
+static loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
+{
+    loff_t res, size;
+    struct aesd_dev *dev = filp->private_data;
+
+    if (!dev) 
+       return -EINVAL;
+
+   
+    if (mutex_lock_interruptible(&dev->lock))
+        return -ERESTARTSYS;
+
+    size = aesd_buffer_size_bytes(&dev->cmd_history); 
+
+   
+    res = fixed_size_llseek(filp, offset, whence, size);
+
+    mutex_unlock(&dev->lock);
+    return res;   /* returns new position */
+}
+
+
+/**
+ * @brief Calculate absolute byte offset for a given write command and offset
+ *
+ * @param circ_buf   Pointer to the AESD circular buffer
+ * @param cmd_index  Command number to seek into (0-based)
+ * @param byte_offset Byte offset inside that command
+ * @param absolute_pos Pointer to store computed absolute position
+ *
```

Comment:
Would be good to note caller must hold lock here.  Also might make sense to move to the circular buffer .c file.